File: Advanced_Robotics_Course/Chapter16/Lesson1/Chapter16_Lesson1.m
T = 0.1;
L = 1.0;
u_max = 0.5;
N_horizon = 10;
dx = 0.1;

controls = [-u_max, 0.0, u_max];
states = -L:dx:L;
initial_states = 0.0;  % could also use a vector of initial states

% Represent visited as logical matrix indexed by (state_index, step)
n_states = numel(states);
visited = false(n_states, N_horizon + 1);

% Map from x to index
[~, idx0] = min(abs(states - initial_states));
queue = [idx0, 0];  % [state_index, step]
visited(idx0, 1) = true;

safe = true;

while ~isempty(queue) && safe
    x_idx = queue(1,1);
    step  = queue(1,2);
    queue(1,:) = [];

    x = states(x_idx);
    if step == N_horizon
        continue;
    end

    for u = controls
        x_next = x + T * u;
        if x_next < -L || x_next > L
            fprintf('Unsafe successor from x=%f with u=%f\n', x, u);
            safe = false;
            break;
        end
        % snap to grid
        [~, idx_next] = min(abs(states - x_next));
        if ~visited(idx_next, step + 2)
            visited(idx_next, step + 2) = true;
            queue(end+1,:) = [idx_next, step + 1]; %#ok<AGROW>
        end
    end
end

fprintf('Abstraction declared SAFE up to horizon %d: %d\n', ...
        N_horizon, safe);
      

File: Advanced_Robotics_Course/Chapter16/Lesson1/Chapter16_Lesson1.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <cmath>
#include <unordered_set>
#include <tuple>

struct StateStepHash {
    std::size_t operator()(const std::pair<double,int>& s) const {
        // Simple hash: combine rounded x and step
        long xi = std::lround(s.first * 100); // 2 decimals
        return std::hash<long>{}(xi) ^ (std::hash<int>{}(s.second) << 1);
    }
};

int main() {
    const double T = 0.1;
    const double L = 1.0;
    const double u_max = 0.5;
    const int N_horizon = 10;
    const double dx = 0.1;

    std::vector<double> controls = {-u_max, 0.0, u_max};
    std::vector<double> initial_states = {0.0};

    std::queue<std::pair<double,int>> q;
    std::unordered_set<std::pair<double,int>, StateStepHash> visited;

    for (double x0 : initial_states) {
        q.push({x0, 0});
        visited.insert({x0, 0});
    }

    auto succ = [&](double x) {
        std::vector<std::pair<double,bool>> result;
        for (double u : controls) {
            double x_next = x + T * u;
            if (x_next <= L && x_next >= -L) {
                double x_disc = dx * std::round(x_next / dx);
                result.push_back({x_disc, true});
            } else {
                result.push_back({0.0, false}); // unsafe
            }
        }
        return result;
    };

    bool safe = true;
    while (!q.empty() && safe) {
        auto [x, step] = q.front();
        q.pop();
        if (step == N_horizon) continue;

        for (auto& pair : succ(x)) {
            double x_next = pair.first;
            bool inside = pair.second;
            if (!inside) {
                std::cout << "Unsafe successor from x=" << x << std::endl;
                safe = false;
                break;
            }
            std::pair<double,int> node{x_next, step + 1};
            if (visited.find(node) == visited.end()) {
                visited.insert(node);
                q.push(node);
            }
        }
    }

    std::cout << "Abstraction declared SAFE up to horizon "
              << N_horizon << ": " << std::boolalpha << safe << std::endl;

    return 0;
}
      
